action/protocol/staking/staking_statereader.go

```diff
@@ -0,0 +1,141 @@
+	return &compositiveStakingStateReader{
+		liquidIndexer: liquidIndexer,
+		nativeIndexer: nativeIndexer,
+		nativeSR:      nativeSR,
+	}
+}
+
+func (c *compositiveStakingStateReader) readStateBuckets(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBuckets) (*iotextypes.VoteBucketList, uint64, error) {
+	// TODO (iip-13): combine native and liquid staking buckets
+	return c.nativeSR.readStateBuckets(ctx, req)
+}
+
+func (c *compositiveStakingStateReader) readStateBucketsByVoter(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBucketsByVoter) (*iotextypes.VoteBucketList, uint64, error) {
+	// TODO (iip-13): combine native and liquid staking buckets
+	return c.nativeSR.readStateBucketsByVoter(ctx, req)
+}
+
+func (c *compositiveStakingStateReader) readStateBucketsByCandidate(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate) (*iotextypes.VoteBucketList, uint64, error) {
+	// TODO (iip-13): combine native and liquid staking buckets
+	return c.nativeSR.readStateBucketsByCandidate(ctx, req)
+}
+
+func (c *compositiveStakingStateReader) readStateBucketByIndices(ctx context.Context, req *iotexapi.ReadStakingDataRequest_VoteBucketsByIndexes) (*iotextypes.VoteBucketList, uint64, error) {
+	// TODO (iip-13): combine native and liquid staking buckets
+	return c.nativeSR.readStateBucketByIndices(ctx, req)
+}
+
+func (c *compositiveStakingStateReader) readStateBucketCount(ctx context.Context, _ *iotexapi.ReadStakingDataRequest_BucketsCount) (*iotextypes.BucketsCount, uint64, error) {
+	// TODO (iip-13): combine native and liquid staking buckets
+	return c.nativeSR.readStateBucketCount(ctx, nil)
+}
+
+func (c *compositiveStakingStateReader) readStateCandidates(ctx context.Context, req *iotexapi.ReadStakingDataRequest_Candidates) (*iotextypes.CandidateListV2, uint64, error) {
+	// get height arg
+	inputHeight, err := c.nativeSR.SR().Height()
+	if err != nil {
+		return nil, 0, err
+	}
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	epochStartHeight := rp.GetEpochHeight(rp.GetEpochNum(inputHeight))
+
+	// read native candidates
+	var (
+		candidates      *iotextypes.CandidateListV2
+		height          uint64
+		candidatesBytes []byte
+	)
+	if epochStartHeight != 0 && c.nativeIndexer != nil {
+		// read candidates from indexer
+		candidatesBytes, height, err = c.nativeIndexer.GetCandidates(height, req.GetPagination().GetOffset(), req.GetPagination().GetLimit())
+		if err != nil {
+			return nil, height, err
+		}
+		candidates = &iotextypes.CandidateListV2{}
+		if err = proto.Unmarshal(candidatesBytes, candidates); err != nil {
+			return nil, height, errors.Wrap(err, "failed to unmarshal candidates")
+		}
+	} else {
+		// read candidates from native state
+		candidates, height, err = c.nativeSR.readStateCandidates(ctx, req)
+		if err != nil {
+			return candidates, height, err
```

Comment:
why is `candidates, height, err` instead of `nil, 0, err` returned?

Reply:
just maintain the original logic in case of native, but which may cause confusing. return `nil, 0, err` should be more clear when error occurs